### DISCUSSION TITLE
DB connections

### DISCUSSION THREAD

**Discussion by thielj** [lldap/lldap#910] [General]
Hi

I've started to play around with a (serverless, SSL secured) Postgres as DB backend. Queries seem to be awfully slow. Is it possible that connections aren't kept open? I'm no Rust programmer, but I was wondering if this might need e.g. min_connections(2) and a decent idle_timeout(60)?

```
        let mut sql_opt = sea_orm::ConnectOptions::new(database_url.to_string());
        sql_opt
            .max_connections(5)
            .sqlx_logging(true)
            .sqlx_logging_level(log::LevelFilter::Debug);
```

https://docs.rs/sqlx/latest/sqlx/struct.Pool.html

My understanding from the above and a bit snooping around the code is that the default "pool" is quite dry, i.e. 

> The pool will establish connections only as needed.

> The default configuration is mainly suited for testing and light-duty applications. For production applications, you’ll likely want to make at least few tweaks.

(I tried a docker build with min_connections and idle_timeout but couldn't get past `error: failed to compile 'wasm-pack v0.12.1'`)

**Comment by nitnelave** [comment]
Well, this _is_ a light-duty application (unless you have a very different use case for lldap than what I had in mind).

But it seems that setting a `min_connection(1)` could do the trick (how many requests in parallel do you usually make?)

**Comment by nitnelave** [comment]
Note that you can also enable debug mode and see for each query what takes time.

**Reply by thielj** [reply]
No need to benchmark. The time is consumed during establishing the connection. The query itself is negligible, usually a single row from the user table, but end up taking more than one second.

I don't have many requests at all at the moment, but I would still configure two connections, just to have one spare. It might not make a difference for a local SQLite DB, but it's huge (insert Trump meme) for a remote server requiring SSL handshake, authentication, etc.

**Reply by nitnelave** [reply]
Makes sense. Do you want to send a PR?

**Reply by thielj** [reply]
As I said, I couldn't build/test it :( any hints?

**Reply by nitnelave** [reply]
Hmm, not sure, since I don't know your setup. Have you tried the lldap/rust_dev image? (I think that's what it's called)

**Reply by thielj** [reply]
I tried a `docker build .` using the images in Dockerfile

```
256.5 error: failed to compile `wasm-pack v0.12.1`, intermediate artifacts can be found at `/tmp/cargo-installFPDn6k`
256.5
256.5 Caused by:
256.5   package `clap_derive v4.5.4` cannot be built because it requires rustc 1.74 or newer, while the currently active rustc version is 1.69.0
256.5   Try re-running cargo install with `--locked`
256.5 error: failed to compile `cargo-chef v0.1.66`, intermediate artifacts can be found at `/tmp/cargo-installBodofi`
256.5
256.5 Caused by:
256.5   package `cargo-platform v0.1.8` cannot be built because it requires rustc 1.73 or newer, while the currently active rustc version is 1.69.0
256.5   Try re-running cargo install with `--locked`
256.5      Summary Failed to install wasm-pack, cargo-chef (see error(s) above).
256.5 error: some crates failed to install
```

Looks like I need a more recent builder, rustc>1.74

**Reply by nitnelave** [reply]
That Dockerfile is not well maintained, it might need updating

**Reply by thielj** [reply]
Hahaha. Definitely. rust:alpine3.18 seems to do the job. Let me test this first and get back to you tomorrow. I'm sure these queries can run faster than 3.7 seconds...

> 🚧 [warn]:  | summary: "SELECT \"users\".\"password_hash\" FROM \"users\" …" | db.statement: "\n\nSELECT\n  \"users\".\"password_hash\"\nFROM\n  \"users\"\nWHERE\n  \"users\".\"user_id\" = $1\nLIMIT\n  $2\n" | rows_affected: 0 | rows_returned: 1 | elapsed: 3.694387658s

> 🚧 [warn]:  | summary: "SELECT \"users\".\"password_hash\" FROM \"users\" …" | db.statement: "\n\nSELECT\n  \"users\".\"password_hash\"\nFROM\n  \"users\"\nWHERE\n  \"users\".\"user_id\" = $1\nLIMIT\n  $2\n" | rows_affected: 0 | rows_returned: 1 | elapsed: 2.221506552s

**Comment by thielj** [comment]
Some findings and suggestions.

## Short version: 

Make it configurable. 

If you don't fancy that, or need defaults, I would suggest min_connections(4 + #cores) for embedded/SQLite and (8 + #cores) for anything else, and max_connections of twice the minimum. Plus sensible timeouts for remote servers to reflect the high connection costs. 

## Long version: 

With the current settings, I was reaching the max_connections(5) very quickly. When I lifted this to max_connections(16), I peaked at 8 quickly. It might need another 1 or 2 when the scheduler kicks in. I'm not sure why it needs all these connections though. I'm certainly not trying to authenticate 8 users in parallel. Connections not being released in time / waiting to be GC'ed, maybe? As I said, I don't know Rust... 

Could it be related to pooled LDAP connections holding on to DB connections? I'm using the Authelia feature build with 4 pooled LDAP connections. This is what I get:

```
# docker exec ldap netstat
Active Internet connections (w/o servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State
tcp        0      0 ldap$.xxx.docker.internal:ldap authelia.services:35750 ESTABLISHED
tcp        0      0 ldap$.xxx.docker.internal:57280 ec2-xxx:postgresql ESTABLISHED
tcp        0      0 ldap$.xxx.docker.internal:60618 ec2-xxx:postgresql TIME_WAIT
tcp        0      0 ldap$.xxx.docker.internal:33144 ec2-xxx:postgresql ESTABLISHED
tcp        0      0 ldap$.xxx.docker.internal:37294 ec2-xxx:postgresql TIME_WAIT
tcp        0      0 ldap$.xxx.docker.internal:58052 ec2-xxx:postgresql ESTABLISHED
tcp        0      0 ldap$.xxx.docker.internal:34336 ec2-xxx:postgresql TIME_WAIT
tcp        0      0 ldap$.xxx.docker.internal:ldap authelia.services:35766 ESTABLISHED
tcp        0      0 ldap$.xxx.docker.internal:ldap authelia.services:35738 ESTABLISHED
tcp        0      0 ldap$.xxx.docker.internal:33920 ec2-xxx:postgresql ESTABLISHED
tcp        0      0 ldap$.xxx.docker.internal:ldap authelia.services:35764 ESTABLISHED
tcp        0      0 ldap$.xxx.docker.internal:38974 ec2-xxx:postgresql TIME_WAIT
```  

In general, I would suggest that at least min/max is configurable. 

It may not matter much for an embedded SQLite instance with minimal overheads, but for an SSL secured server, consider: it's 3 round trips to establish the SSL connection, at least one round-trip to authenticate with the DB and another for the actual query. Add to that processing, crypto, hashing, load balancers, etc. It was enough for LLDAP to show warnings when simple queries took longer than one second...

We don't need to optimize for throughput and LLDAP is unlikely to saturate the DB engine (and if it does, it's time to upgrade). When optimizing for latency, I want to make sure that no query has to wait for a connection to be established or returned to the pool. The resource overhead for keeping an established connection around is neglectable. So I rather have a few too many connections, unless there are limits server side that would starve other apps.

---

The below works well for me, for example. Connections would time-out after 5 minutes and be forcefully recycled each hour, just in case. I think PgBouncer uses similar timeouts.

```
            .min_connections(8)
            .max_connections(16)
            .idle_timeout( Duration::from_secs(300) )
            .max_lifetime( Duration::from_secs(3600) )
```

**Reply by nitnelave** [reply]
Sounds like you did your homework 😅
Yeah, making it configurable (maybe under an advanced section?) works for me.

**Reply by thielj** [reply]
Everything to never ever touch OpenLDAP again ;)

**Reply by nitnelave** [reply]
Do you want to send a PR with this?

**Comment by thielj** [comment]
Naa. These "docker build" runs take forever. Adding more than those 3 lines
would require setting up a Rust dev environment first.

**Reply by nitnelave** [reply]
That's not too hard. rustup gets you the compiler, cargo and the tools, and then it's just a `cargo build` and you have all the dependencies. `cargo test` to run the tests, `cargo fmt` to reformat the code.

**Reply by nitnelave** [reply]
I'm saying that because I don't have much time behind a computer to write code, so it probably won't be done for a long time unless you do it. But I can help with questions and reviews.

**Comment by thielj** [comment]
That's okay with me. I have a working image now and can share it here if
someone else runs into this issue.

Learning Rust is on my list - but it won't be this month or next and
probably not this year either 😆😭

**Comment by thielj** [comment]
Let's revisit this when Authelia releases the LDAP pooling and I update my
own setup.

**Comment by thielj** [comment]
@nitnelave Looks like these didn't make it into the new release?

**Reply by nitnelave** [reply]
No, I don't think there was a PR about it.

**Reply by thielj** [reply]
I can do a PR for the 4 lines mentioned above. They wouldn't hurt anyone, but people with bigger machines (and, supposedly, more RAM and more users) would still benefit from higher defaults. 

```
            .min_connections(8)
            .max_connections(16)
            .idle_timeout( Duration::from_secs(300) )
            .max_lifetime( Duration::from_secs(3600) )
```

Sorry, I'm not a Rust developer. I wouldn't know how to implement a static config or a dynamic config based on CPU cores, unless I ask AI to have a go at this.

**Reply by nitnelave** [reply]
Let's keep it at 8/16 for now, especially given that we have a single worker thread (so the number of CPUs doesn't matter)

Could you send a PR with that, with a summary of your analysis?

Thanks!

**Comment by thielj** [comment]
@nitnelave Is there a particular reason for the single LDAP worker thread?

**Reply by nitnelave** [reply]
No one has ever shown me a workload that would justify even a second thread. 

It's async, so there can be several in-flight requests at the same time, it's just the cpu-bound part that gets serialized

**Reply by thielj** [reply]
Which means that during Argon2 hashing (bind/login), all other activity is blocked for a significant period, right?

**Comment by nitnelave** [comment]
Yes, but that's usually about half a second or less

**Reply by thielj** [reply]
Exactly. Those 500ms are stalling requests that would otherwise complete in less than 0.1ms.

**Reply by nitnelave** [reply]
Right, but is that an actual problem with an actual workload that you have? Can you give me some numbers in terms of number of users, requests/s, p50/90/99?

Or are you just worried about the occasional page load slowdown when another user is logging in at exactly the same moment?

**Comment by thielj** [comment]
My primary concern isn't about typical user login patterns, where
individuals log in infrequently and then rely on cookies for subsequent
access. Instead, the issue arises with services that utilize
`Authorization:` headers, as each access triggers a lengthy password
verification process. This, in turn, causes significant stalling for all
other incoming requests.

Given this scenario, I'm wondering why we couldn't enable multiple workers.
As I understand it, each additional thread would consume only a few KB of
RAM, which seems like a negligible overhead compared to the benefits of
improved concurrency.

**Reply by nitnelave** [reply]
Again, unless you have a significant load (and I'd be interested in latency numbers), it seems that the better solution is to cache the result of the authorization headers, maybe for even as short as one minute (but 5 min should be fine). Otherwise you'll still pay the password check for every request.

**Reply by thielj** [reply]
I don't think Authelia can do this, but I'll have a look. 

Anyway, the services are not the problem, they can wait a few hundred milliseconds. Just run ( 4 + #cores ) worker threads, put a mutex or semaphore around the hashing (to prevent excessive n*50MB memory spikes) and you should be good. I'll have a look myself next week.

Have you considered adding https://rqlite.io/ support? I can give this a try next week, if that's something you would be interested in?

**Reply by nitnelave** [reply]
You still haven't convinced me of the utility of this feature, I'm not likely to accept the PR. 

As for rqlite, our DB support relies entirely on sqlx, so that is probably more effort than you want to put in.

**Reply by nitnelave** [reply]
See https://github.com/authelia/authelia/issues/5006

AFAIU, authelia can generate bearer tokens that have a certain lifetime, this "caching" the authentication result.

**Reply by thielj** [reply]
Looks like Authelia has added caching now. LLDAP would benefit nonetheless, but I'm tired of "convincing" people.

Re rqlite, does this look like too much effort: https://github.com/HaHa421/sqlx-rqlite
